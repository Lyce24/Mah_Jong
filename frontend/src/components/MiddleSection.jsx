import Discard from "./Discard";
import RecentDiscard from "./RecentDiscard";
import "../index.css";

export default function MiddleSection({ discardPile }) {
  const usernames = Array(4).fill(null);

  return (
    <div className="middleSection">
      <Discard
        orientation="playerDiscard"
        discardPile={discardPile}
        username={usernames[0]}
      />
      <br />
      <Discard
        orientation="leftDiscard"
        discardPile={discardPile}
        username={usernames[1]}
      />
      <br />
      <Discard
        orientation="rightDiscard"
        discardPile={discardPile}
        username={usernames[2]}
      />
      <br />
      <Discard
        orientation="topDiscard"
        discardPile={discardPile}
        username={usernames[3]}
      />
      <br />
      <RecentDiscard />
    </div>
  );
}
