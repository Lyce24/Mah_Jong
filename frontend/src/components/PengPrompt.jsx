import "../index.css";

export default function PengPrompt(props) {
  return (
    <div className="pengPrompt">
      <button type="button" onClick={props.onClick} className="button">
        Peng?
      </button>
    </div>
  );
}
