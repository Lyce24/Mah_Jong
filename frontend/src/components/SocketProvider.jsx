import { w3cwebsocket as W3CWebSocket } from "websocket";
import React, { createContext, useContext, useEffect, useState } from "react";
import { useUsername } from "./UsernameProvider";

// create SocketContext with value = the socket instance
// so all components can read this value with useContext(SocketContext)
const SocketContext = createContext();

// useSocket is a custom hook
// we can use it in any component to get the socket instance
export const useSocket = () => useContext(SocketContext);

// SocketProvider is a component that wraps all other components
// so all components can access the context value, which is the socket instance
export const SocketProvider = ({ children }) => {
  const URL = "ws://localhost:8000/ws/socket-server";
  const [socket, setSocket] = useState(null);

  const username = useUsername();

  useEffect(() => {
    // initialize socket instance
    const newSocket = new WebSocketInstance(URL, username);
    setSocket(newSocket);

    // this is the cleanup function, it will be called when the component unmounts (ideally never)
    return () => newSocket.disconnect();
  }, [username]);

  return (
    <SocketContext.Provider value={socket}>
      {socket && children}
      {/* conditionally render the child components only when the socket state is not null. */}
    </SocketContext.Provider>
  );
};

// Create cleaner interface for websocket interface interaction on frontend
class WebSocketInstance {
  constructor(URL, username) {
    this.socketRef = new W3CWebSocket(URL);
    this.readyState = this.socketRef.readyState;
    this.username = username;
    this.addCallbacks();
    this.addListener();
  }

  // never used
  connect() {
    this.socketRef.onopen = () => {
      console.log("WebSocket connected");
    };
  }

  // never used
  disconnect() {
    this.socketRef.close();
  }

  // used once during initialization
  addCallbacks() {
    this.socketRef.onerror = (e) => {
      console.log(e.message);
    };

    this.socketRef.onclose = () => {
      console.log("WebSocket closed");
    };
  }

  // Send JSON object to backend
  send(message) {
    if (this.socketRef.readyState === WebSocket.OPEN) {
      // Append username to the message JSON being sent
      const newmessage = {
        ...message,
        username: this.username,
      };
      console.log("send message", newmessage);
      this.socketRef.send(JSON.stringify(newmessage));
    } else {
      console.error("Socket is not connected");
    }
  }

  /* Backend response JSON structure: 
  status: "202" or "400"
  result_type: "room_id", "start_tiles", "draw_tile" etc
  */
  // Add default listener to display messages from backend in console
  addListener() {
    this.socketRef.onmessage = function (e) {
      if (typeof e.data === "string") {
        const message = JSON.parse(e.data);
        console.log("Received from general listener: ", message);
      }
    };
  }

  // Add listner to set room id and navigate to room page
  // Ideally, it would be added only once, when user clicks create room button or the join room button
  addRoomListener(setRoomNum, navigate) {
    this.socketRef.onmessage = function (e) {
      if (typeof e.data === "string") {
        const message = JSON.parse(e.data);
        console.log("Received from room listener: ", message);
        // only proceed if message is about create room
        if (message.result_type === "room_id") {
          console.log("Received: ", message);
          if (message.status !== "202") {
            console.log("create room error");
            return;
          }
          setRoomNum(message.room_id);
          navigate(`/room/${message.room_id}`);
        }
      }
    };
  }

  addPlayerListener(setHand, setDrawnTile) {
    this.socketRef.onmessage = function (e) {
      if (typeof e.data === "string") {
        const message = JSON.parse(e.data);
        console.log("Received from player listener: ", message);
        // only proceed if message is for this player, and message is successful
        if (message.player !== this.username) {
          return;
        } else if (message.status !== "202") {
          console.log("Received error message", message);
          return;
        }
        console.log("Received: ", message);
        switch (message.result_type) {
          case "start_tiles":
            setHand(message.tiles);
        }
      }
    };
  }

  // Add listener to set tile that has been drawn
  addDrawListener(setTile) {
    this.socketRef.onmessage = function (e) {
      if (typeof e.data === "string") {
        const message = JSON.parse(e.data);
        console.log("Received event from draw listener: ", message);
        // only proceed if message if for this player
        if (message.player !== this.username) {
          return;
        }
        // only proceed if message is about draw
        else if (message.result_type === "draw_tile") {
          console.log("Received: ", message);
          if (message.status !== "202") {
            setTile(null);
            return;
          }
          // Tile has suite, number, index, key params
          // Backend should send suite and number
          // TODO: generate unique key for tile, use null for index, append to backend response
          /* const newmessage = {
            ...message.tile,
            index: null,
            key: nanoid(),
          }; */
          setTile(message.tile);
        }
      }
    };
  }

  // Add listener to display tile that has been discarded (by any player)
  addDiscardListener(setDiscardPile) {
    this.socketRef.onmessage = function (e) {
      if (typeof e.data === "string") {
        const message = JSON.parse(e.data);
        console.log("Received event from discard listener: ", message);
        // only proceed if message is about discard
        if (message.result_type === "discard_tile") {
          console.log("Received: ", message);
          if (message.status !== "202") {
            console.log("discard error");
            return;
          }
          setDiscardPile(message.tile);
        }
      }
    };
  }

  // Abandoned code
  // Don't send anything, just specify what behavior you want when receive backend response
  receive(setResult) {
    this.socketRef.onmessage = function (e) {
      if (typeof e.data === "string") {
        const message = JSON.parse(e.data);
        console.log("Received: ", message);
        if (message.status !== "202") {
          setResult(null);
          return;
        }
        switch (message.result) {
          case "tile":
            setResult(message.tile);
            break;
          default:
            break;
        }
      }
    };
  }
}
