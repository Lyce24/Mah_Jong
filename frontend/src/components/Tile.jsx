import { useState } from 'react';
import img from '../tiles/back.png';
import * as images from '../tiles/back.png';

export default function Tile(props) {

    const [clickState, setState] = useState(false);

    function handleClick() {
        // toggle clickState
        setState(!clickState);
        console.log(!clickState);
    }
    let clicked = `${clickState ? "clicked" : ""}`;

    if (props.isFacedDown === "true") {
        return (
            <div className={`tile faceDown ${clicked}`} onClick={handleClick}>
                <img src={img} alt={`Facedown`} />
            </div>
        );
    } else {
        let suite = props.suite;
        let number = props.number;
        return (
            <div className={`tile ${clicked}`} onClick={handleClick}>
                <img src={`./images/tile-${suite}-${number}`} alt={`${suite} ${number}`} />
            </div>
        );
    }
}