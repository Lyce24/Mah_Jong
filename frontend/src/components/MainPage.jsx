import CreateRoomButton from './CreateRoomButton';
import JoinRoom from './JoinRoom';
//import WebSocketInstance from './WebSocketInstance';
import { w3cwebsocket as W3CWebSocket } from 'websocket';
import { useState, useEffect } from 'react';
import { useSocket } from './SocketProvider';

export default function MainPage() {
        
    const socket = useSocket();
    //console.log('state', socket.state());
    //setTimeout(()=>{console.log('state', socket.state());}, 1000)
    socket.onopen = () =>{
        console.log("WebSocket Client Connected");
        socket.send(JSON.stringify({
            'type': 'placeholder',
            'message': 'hi',
        }))
    }
    socket.onmessage = function(e) {
        if (typeof e.data === 'string') {
            const message = JSON.parse(`${e.data}`);
            console.log("Received: ", message);
        }
    };

    return (
        <div className='page mainPage'>
            <h3>Welcome to Reed Mahjong!</h3>
            <hr></hr>
            <CreateRoomButton />
            <br />
            <JoinRoom />
        </div>
    );
}