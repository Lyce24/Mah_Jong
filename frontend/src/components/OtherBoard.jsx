import Tile from './Tile';
import { useState } from 'react';
import '../index.css';

export default function OtherBoard() {

    let initialTiles = Array(3).fill({
        suite: "bamboo",
        number: 1
    })
    const [hand, setHand] = useState(initialTiles);

    return (
        <div className='playerBoard otherBoard'>
            {hand.map(tile => (
                <Tile suite={tile.suite} number={tile.number} isFacedDown="true"/>
            ))}
        </div>
    );
}