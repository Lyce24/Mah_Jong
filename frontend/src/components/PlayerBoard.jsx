import Tile from './Tile';
import { useState, useRef } from 'react';
import '../index.css';
import DiscardButton from './DiscardButton';

export default function PlayerBoard() {

    let reference = useRef(0);

    // 13 tiles + 1 drawn tile
    let initialTiles = Array(13);
    for (let i = 0; i<13; i++) {
        initialTiles.push({
        suite: "bamboo",
        number: Math.min(i+1,9),
        index: i
        })
    }
    const [hand, setHand] = useState(initialTiles);
    const [selectedTile, setSelectedTile] = useState(null);

    function handleTileClick(index) {
        if (index === selectedTile) {
            // If the clicked child is already selected, deselect it
            setSelectedTile(null);
        } else {
            // Otherwise, deselect the previously selected child and select the new child
            setSelectedTile(index);
        }
    }
    //const [drawnTile, setTile] = useState({suite: "", number: 0});

    /* function handleDiscard(params) {
        console.log(reference.current.state)
        // collect tile that is clicked
    } */
    
    // Template for adding elements to array
    /* const temp = () => {
        let updateHand = [...hand, {
            suite: "bamboo",
            number: 1
        }];
        setHand(updateHand);
    }; */
    
//<DiscardButton onClick={handleDiscard}/>
    return (
        <div className='board playerBoard'>
            {hand.map(tile => (
                <Tile suite={tile.suite} number={tile.number} index={tile.index} onClick={handleTileClick} isSelected={selectedTile == tile.index} isFacedDown="false"/>
            ))}
        </div>
        
    );
}