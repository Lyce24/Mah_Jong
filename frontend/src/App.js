import MainPage from './components/MainPage';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import RoomPage from './components/RoomPage';

export default function App(prop) {
    return (
      <BrowserRouter>
        <Routes>
          <Route path='/' element={<MainPage />}/>
          <Route path='/room/:roomid' element={<RoomPage/>}/>
        </Routes>
      </BrowserRouter>
      
    );
}
