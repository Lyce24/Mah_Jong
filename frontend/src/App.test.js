
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { UsernameProvider } from './components/UsernameProvider';
import { SocketProvider } from './components/SocketProvider';
import { BrowserRouter, Router } from 'react-router-dom';
import * as TestRenderer from "react-test-renderer";
import { waitFor } from '@testing-library/react';
//import { useSocket } from "./SocketProvider";
import { createMemoryHistory } from 'history';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import renderer from 'react-test-renderer';

import App from './App';
import MainPage from './components/MainPage';
import RoomPage from './components/RoomPage';
import CreateRoomButton from './components/CreateRoomButton';

// test app renders correctly with context
xdescribe('App component', () => {
  it('app renders without crashing', () => {
    render(<App />);
  });
  it('app renders main page', async () => {
    render(<App />);
    expect(screen.getByRole('heading')).toHaveTextContent('Welcome to Reed Mahjong!')
    const createGameButton = screen.getByText('Create a Game!');
    expect(createGameButton).toBeInTheDocument();
    const joinRoomButton = screen.getByText('Submit');
    expect(joinRoomButton).toBeInTheDocument();
  });
  xit('renders app without crashing with snapshot', () => {
    const tree = renderer
      .create(<App />)
      .toJSON();
  expect(tree).toMatchSnapshot();
  });
});

/* // mock useNavigate hook
const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

// mock useNavigate hook
const mockedUsedSocket = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useSocket: () => mockedUsedSocket,
}));

// Create a mock implementation of the WebSocket instance
const mockWebSocket = {
  send: jest.fn(),
  addRoomListener: jest.fn(),
  close: jest.fn(),
};

// Mock the implementation of the SocketContext provider
jest.mock('./SocketProvider', () => ({
  SocketProvider: {
    Consumer: ({ children }) => children(mockWebSocket),
  },
}));

test('useSocket returns the correct socket instance', () => {
  const { result } = renderHook(() => useSocket());
  expect(result.current).toBe(mockWebSocket);
}); */

/* jest.mock('./components/SocketProvider', () => ({
  useSocket: jest.fn(),
})); */

xdescribe('MainPage Mock', () => {
  test('renders main page correctly', () => {
    const mockSocket = {
      addRoomListener: jest.fn(),
    };
    const mockNavigate = jest.fn();
    const mockSetRoomNum = jest.fn();

    // Mock the useSocket hook to return the mock socket instance
    jest.spyOn(require('./components/SocketProvider'), 'useSocket').mockReturnValue(mockSocket);
    jest.spyOn(require('react-router-dom'), 'useNavigate').mockReturnValue(mockNavigate);

    render(<MainPage navigate={mockNavigate} setRoomNum={mockSetRoomNum} />);

    // Assert that the page header is rendered
    expect(screen.getByRole('heading', { name: /Welcome to Reed Mahjong!/i })).toBeInTheDocument();

    // Assert that the CreateRoomButton component is rendered
    expect(screen.getByRole('button', { name: /Create Room/i })).toBeInTheDocument();

    // Assert that the JoinRoom component is rendered
    expect(screen.getByRole('button', { name: /Join Room/i })).toBeInTheDocument();

    // Assert that the socket.addRoomListener function is called with the expected arguments
    expect(mockSocket.addRoomListener).toHaveBeenCalledWith(mockSetRoomNum, mockNavigate);
  });
});

// wrapper function to render components with context
function renderWithinApp(component) {
  return render(
    <App>{component}</App>
  );
}

function renderWithContext(component){
  render(
    <UsernameProvider>
      <SocketProvider>
        <BrowserRouter>
          {component}  
        </BrowserRouter>
      </SocketProvider>
    </UsernameProvider>
  )
}

function renderWithURL(route) {
  //const history = createMemoryHistory({ initialEntries: [route] });

  /* let renderer: TestRenderer.ReactTestRenderer;
  TestRenderer.act(() => {
    renderer = TestRenderer.create(
      <UsernameProvider>
        <SocketProvider>
          <MemoryRouter initialEntries={["/"]}>
            <Routes>
              <Route path='/' element={<MainPage />}/>
              <Route path='/room/:roomid' element={<RoomPage/>}/>
            </Routes>
          </MemoryRouter>
        </SocketProvider>
      </UsernameProvider>
    )}); */

    //return renderer;


  return render(
    <UsernameProvider>
      <SocketProvider>
        <MemoryRouter initialEntries={[route]}>
          <Routes>
            <Route path='/' element={<MainPage />}/>
            <Route path='/room/:roomid' element={<RoomPage/>}/>
          </Routes>
        </MemoryRouter>
      </SocketProvider>
    </UsernameProvider>
  );
};

xdescribe("<Navigate>", () => {
  describe("with an absolute href", () => {
    xit("navigates to the correct test URL", () => {
      let renderer: TestRenderer.ReactTestRenderer;
      TestRenderer.act(() => {
        renderer = TestRenderer.create(
          <UsernameProvider>
            <SocketProvider>
            <MemoryRouter initialEntries={["/test"]}>
            <Routes>
              <Route path='/' element={<MainPage />}/>
              <Route path='/test' element={<h1>About</h1>}/>
              <Route path='/room/:roomid' element={<RoomPage/>}/>
            </Routes>
          </MemoryRouter>
            </SocketProvider>
          </UsernameProvider>
          
        );
      });

      expect(renderer.toJSON()).toMatchInlineSnapshot(`
        <h1>
          About
        </h1>
      `);
    });

    it("navigates to the correct main URL", () => {
      let renderer: TestRenderer.ReactTestRenderer;
      TestRenderer.act(() => {
        renderer = TestRenderer.create(
          <UsernameProvider>
            <SocketProvider>
              <MemoryRouter initialEntries={["/"]}>
                <Routes>
                  <Route path='/' element={<MainPage />}/>
                  <Route path='/test' element={<h1>About</h1>}/>
                  <Route path='/room/:roomid' element={<RoomPage/>}/>
                </Routes>
              </MemoryRouter>
            </SocketProvider>
          </UsernameProvider>
          
        );
      });


      expect(renderer.root.findByProps({className: 'createRoomButton'}).children[0].props.type).toBe('button');
      //console.log(renderer.root.findAllByType(CreateRoomButton).type);
      //console.log(renderer.root.findByType('div').children);
      console.log(renderer.toJSON())
    });
  });
})

xdescribe('Main Page component', () => {
  it('renders correctly with context and router', async () => {
    renderWithContext(<MainPage />);

    expect(screen.getByRole('heading',)).toHaveTextContent('Welcome to Reed Mahjong!')
    const createGameButton = screen.getByText('Create a Game!');
    expect(createGameButton).toBeInTheDocument();
    const joinRoomButton = screen.getByText('Submit');
    expect(joinRoomButton).toBeInTheDocument();
    expect(location.pathname).toBe('/');
  });

  xit('renders correctly with mock useNavigate', async () => {
    render(<MainPage />);

    expect(screen.getByRole('heading', {level: 3})).toHaveTextContent('Welcome to Reed Mahjong!')
    const createGameButton = screen.getByText('Create a Game!');
    expect(createGameButton).toBeInTheDocument();
    const joinRoomButton = screen.getByText('Submit');
    expect(joinRoomButton).toBeInTheDocument();
  });

  xit('click create game button', async () => {
    
    //const history = createMemoryHistory(); // create a mock history object
    render(
        <MainPage />);
    const createGameButton = screen.getByText('Create a Game!');
    expect(createGameButton).toBeInTheDocument();
    await fireEvent.click(createGameButton);
    await waitFor(() => {
      expect(history.location.pathname).toMatch(/\/room\/\d+/);
      //const roomHeading = getByText('Room'); // get the Room heading
      //expect(roomHeading).toBeInTheDocument(); // check that the Room heading is rendered
    }, { timeout: 5000 }); // increase the timeout if needed
    //expect(screen.getByRole('heading')).toHaveTextContent('Room');
  });
});

describe('Room Page component', () => {
  it('renders without crashing, null room id', async () => {
    renderWithContext(<RoomPage />);
    expect(screen.getByRole('heading', {level: 1})).toHaveTextContent('Room');
  });

  it('renders with room id 123', async () => {
    //renderWithContext(<RoomPage />);
    const roomid = 123;
    const { container } = renderWithURL('/room/123');
  
    console.log('current URL👉️', window.location.href);
    console.log('URL', location.pathname);
    expect(screen.getByRole('heading', {level: 1})).toHaveTextContent('Room 123');
    //await waitFor(() => {screen.getByRole('heading', {level: 1})}, { timeout: 5000 });
    expect(container.querySelector('h1')).toHaveTextContent(`Room ${roomid}`);
  });
});

xtest('loads and displays greeting', async () => {
  // ARRANGE
  //render(<Fetch url="/room/000/" />)

  // ACT
  //await userEvent.click(screen.getByText('Load Greeting'))
  await screen.findByRole('heading')

  // ASSERT
  expect(screen.getByRole('heading')).toHaveTextContent('Welcome')
  //expect(screen.getByRole('button')).toBeDisabled()
})