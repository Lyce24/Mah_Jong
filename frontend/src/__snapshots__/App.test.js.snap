// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigate> with an absolute href navigates to the correct URL 1`] = `
<div
  className="page mainPage"
>
  <h3>
    Welcome to Reed Mahjong!
  </h3>
  <hr />
  <div
    className="createRoomButton"
  >
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      Create a Game!
    </button>
  </div>
  <br />
  <div
    className="joinRoom"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        <span>
          Room Code: 
        </span>
        <input
          maxLength={8}
          minLength={8}
          onChange={[Function]}
          pattern="[0-9]{8}"
          type="text"
          value={0}
        />
      </label>
      <input
        className="button"
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</div>
`;
